packages/tooltip/src/Tooltip/utils/createTooltipTriggerEventHandlers.ts

```diff
@@ -0,0 +1,242 @@
+export function createTooltipTriggerEventHandlers(
+  args: CreateTooltipEventsBaseArgs & CreateTooltipEventsArgsClick,
+): TooltipClickEvents;
+export function createTooltipTriggerEventHandlers(
+  args: CreateTooltipEventsBaseArgs &
+    (CreateTooltipEventsArgsHover | CreateTooltipEventsArgsClick),
+): TooltipHoverEvents | TooltipClickEvents {
+  const { setState, triggerEvent, tooltipRef, isEnabled = true } = args;
+
+  switch (triggerEvent) {
+    case TriggerEvent.Hover:
+      return {
+        onMouseEnter: debounce((e: MouseEvent<HTMLElement>) => {
+          if (isEnabled) {
+            args.onMouseEnter?.(e);
+            // Without this the tooltip sometimes opens without a transition. flushSync prevents this state update from automatically batching. Instead updates are made synchronously.
+            // https://react.dev/reference/react-dom/flushSync#flushing-updates-for-third-party-integrations
+            flushSync(() => {
+              setState(true);
+            });
+          }
+        }, 35),
+        onMouseLeave: debounce((e: MouseEvent<HTMLElement>) => {
+          if (isEnabled) {
+            args.onMouseLeave?.(e);
+            setState(false);
+          }
+        }, 35),
+        onFocus: (e: FocusEvent<HTMLElement>) => {
+          if (isEnabled) {
+            args.onFocus?.(e);
+            setState(true);
+          }
+        },
+        onBlur: (e: FocusEvent<HTMLElement>) => {
+          if (isEnabled) {
+            args.onBlur?.(e);
+            setState(false);
+          }
+        },
+      };
+    case TriggerEvent.Click:
+    default:
+      return {
+        onClick: (e: MouseEvent<HTMLElement>) => {
+          if (isEnabled) {
+            // ensure that we don't close the tooltip when content inside tooltip is clicked
+            if (e.target !== tooltipRef?.current) {
+              args.onClick?.(e);
+              setState(curr => !curr);
+            }
+          }
+        },
+      };
+  }
+}
+
+export function useTooltipTriggerEventHandlers(
+  args: CreateTooltipEventsBaseArgs &
+    (CreateTooltipEventsArgsHover | CreateTooltipEventsArgsClick),
+): {
+  onMouseEnter?: MouseEventHandler<HTMLElement>;
+  onMouseLeave?: MouseEventHandler<HTMLElement>;
+  onFocus?: FocusEventHandler<HTMLElement>;
+  onBlur?: FocusEventHandler<HTMLElement>;
+  onClick?: MouseEventHandler<HTMLElement>;
+} {
+  const { setState, triggerEvent, tooltipRef, isEnabled = true } = args;
+  const isHoverTrigger = triggerEvent === TriggerEvent.Hover;
+  const isClickTrigger = triggerEvent === TriggerEvent.Click;
+
+  // Ensure we have a stable way
+  // of checking if the tooltip is enabled from within the callbacks
+  // without creating a new callback on each render
+  const isEnabledRef = useRef(isEnabled);
+  useEffect(() => {
+    isEnabledRef.current = isEnabled;
+  }, [isEnabled]);
+  const getIsEnabled = useCallback(() => isEnabledRef.current, []);
+
+  return {
+    onMouseEnter: useMemo(
```

Comment:
Could we extract all of these triggers and use them in both `createTooltipTriggerEventHandlers` and `useTooltipTriggerEventHandlers`?